mlir/lib/Transforms/ViewOpGraph.cpp

```diff
@@ -282,6 +294,17 @@ class PrintOpPass : public impl::ViewOpGraphBase<PrintOpPass> {
         prevNode = nextNode;
       }
 
+      if (!IsModuleBlock
+           && onlyEntryAndExitOperations && (&block.front() != &block.back())) {
+        Node nextNode = processOperation(&block.back());
+        if (printControlFlowEdges) {
+          emitEdgeStmt(*prevNode, nextNode, /*label=*/"",
+                       kLineStyleControlFlow);
+        }
+
+        prevNode = nextNode;
+      }
+
```

Comment:
All the above is kinda fragmented, and it's hard to thing about corner cases.
I propose to just take the `if (!IsModuleBlock ...` parts before the loop.

Something along these lines:
```c++
if (!IsModuleBlock and onlyEntryAndExitOperations and block.size() > 2) {

  auto FirstOp = block.front();
  processOperation(FirstOp);
  if (printControlFlowEdges)
    emitEdgeStmt(nullopt, FirstOp, "", kLineStyleControlFlow);
  
  auto LastOp = block.back();
  processOperation(LastOp);
  if (printControlFlowEdges)
    emitEdgeStmt(LastOp, nullopt, "", kLineStyleControlFlow);

} else {
  for (Operation &op : block) {
    // as before
  }
}
```

So the loop just keeps taking care of the case when the option is inactive. When the option is active you just avoid the loop altogether and just print the first and last.

Actually I'm not sure about the two `if (printControlFlowEdges)`, if they are necessary, and if the arguments I'm passing to `emitEdgeStmt` are correct. You might want to double check that.

Reply:
Factoring out the handling of first and last node only seems a very reasonable way to disentangle a bit the code. Done it in 5c0cd71bdc951b02ec75ebf41bd60821825939b5.